src/theory/quantifiers/sygus/sygus_grammar_norm.cpp

```diff
@@ -599,7 +599,7 @@ TypeNode SygusGrammarNorm::normalizeSygusRec(TypeNode tn)
     return tn;
   }
   /* Collect all operators for normalization */
-  const Datatype& dt = tn.getDatatype();
+  const Datatype& dt = static_cast<DatatypeType>(tn.toType()).getDatatype();
```

Comment:
Strictly speaking, this is undefined behavior because `tn.toType()` is not of type `DatatypeType` (UBSan should complain about this). The correct way is:

```suggestion
  const Datatype& dt = DatatypeType(tn.toType()).getDatatype();
```

Reply:
It seems our coding guidelines say to use `static_cast` (https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#other-c-features). Everywhere a cast of this form is done, I ensure that `tn.toType()` is of `DatatypeType`.

Reply:
`TypeNode::toType()` returns an object of type `Type` (i.e. it is not a `DatatypeType` anymore even if the original was because it has been sliced). The cast would be ok if it were on a `Type*` (`static_cast<DatatypeType*>`). I am suggesting to use the `DatatypeType` constructor for the "cast" because that's how our `Type` hierarchy is designed (I don't necessarily agree with the design but that's how it is). See also PR #3420 where I fixed existing casts because UBSan complained.

Reply:
I see, this is indeed confusing.  It's annoying that there is an ambiguity between cast/constructor.  
In light of this, I agree with your changes; will update.